service/service_parameter.go

```diff
@@ -8,25 +8,29 @@ import (
 )
 
 // validateServiceParameters validates data to see if it matches with parameters schema.
-func validateServiceParameters(parameters []*Service_Parameter, data *types.Struct) error {
+func validateServiceParameters(parameters []*Service_Parameter, data []*types.Value) error {
 	var errs xerrors.Errors
-
-	for _, p := range parameters {
+	for i, p := range parameters {
 		var value *types.Value
-		if data != nil && data.Fields != nil {
-			value = data.Fields[p.Key]
+		if data != nil && i < len(data) {
+			value = data[i]
 		}
 		if err := p.Validate(value); err != nil {
 			errs = append(errs, err)
 		}
 	}
-
 	return errs.ErrorOrNil()
 }
 
 // Validate checks if service parameter hash proper types for arrays and objects.
 func (p *Service_Parameter) Validate(value *types.Value) error {
+	isNilOrNullValue := false
 	if value == nil {
+		isNilOrNullValue = true
+	} else {
+		_, isNilOrNullValue = value.GetKind().(*types.Value_NullValue)
+	}
```

Comment:
_, isNilOrNullValue := value.GetKind().(*types.Value_NullValue)
isNilOrNullValue = isNilOrNullValue || value == nil

Reply:
` || value == nil` is useless because it will panic first if value is `nil` in that case